File: pe-math/src/main/java/org/saar/maths/noise/LayeredNoise1f.java
package org.saar.maths.noise;

public class LayeredNoise1f implements Noise1f {

    private final Noise1f noise1f;
    private final int layers;

    public LayeredNoise1f(Noise1f noise1f, int layers) {
        this.noise1f = noise1f;
        this.layers = layers;
    }

    @Override
    public float noise(float x) {
        float noise = 0f;
        for (int i = 0; i < this.layers; i++) {
            final float pow2 = (float) Math.pow(2, i);
            noise += this.noise1f.noise(x / pow2) * pow2;
        }
        return (float) (noise / (Math.pow(2, this.layers) - 1));
    }
}

File: pe-math/src/main/kotlin/org/saar/maths/noise/LayeredNoise1f.kt
package org.saar.maths.noise

import kotlin.math.pow

class LayeredNoise1f(private val noise1f: Noise1f, private val layers: Int) : Noise1f {

    override fun noise(x: Float): Float {
        val noise = (0 until this.layers).sumOf {
            val pow2 = 2.0.pow(it.toDouble()).toFloat()
            (this.noise1f.noise(x / pow2) * pow2).toDouble()
        }
        return (noise / (2.0.pow(this.layers.toDouble()) - 1)).toFloat()
    }
}
